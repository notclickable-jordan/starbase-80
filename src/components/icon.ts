import { is } from "../shared/is";
import { IconAspect } from "../shared/types";
import { Anchor } from "./anchor";

const iconColors = [
	"blue",
	"rose",
	"green",
	"red",
	"yellow",
	"cyan",
	"pink",
	"orange",
	"sky",
	"slate",
	"emerald",
	"zinc",
	"neutral",
	"amber",
	"violet",
];
const iconLevel = 300;

const getIconColor = (index: number) => `bg-${iconColors[iconColors.length % index]}-${iconLevel}`;

interface IProps {
	name: string;
	index: number;
	icon?: string;
	iconColor?: string;
	iconBG?: string;
	iconBubble?: boolean;
	iconBubblePadding?: boolean;
	iconAspect?: IconAspect;
	uri?: string;
	newWindow?: boolean;
	categoryBubblePadding?: boolean;
}

export const Icon = function (props: IProps): string {
	const {
		name,
		uri,
		icon,
		index,
		iconBG,
		iconBubble,
		iconBubblePadding,
		iconColor,
		iconAspect,
		newWindow,
		categoryBubblePadding,
	} = props;

	if (is.null(icon)) {
		if (!is.null(uri)) {
			return Anchor({ uri: uri as string, title: name, newWindow, children: IconBlank({ index }) });
		}

		return IconBlank({ index });
	}

	let bubblePadding = categoryBubblePadding || false;

	if (iconBubblePadding === true) {
		bubblePadding = true;
	} else if (iconBubblePadding === false) {
		bubblePadding = false;
	}

	return IconBase({
		icon: icon as string,
		iconBG,
		iconColor,
		iconBubble,
		iconBubblePadding: bubblePadding,
		iconAspect,
	});
};

interface IIconBlankProps {
	index: number;
}

function IconBlank(props: IIconBlankProps) {
	const { index } = props;
	return `
		<span
			class="${`block w-16 h-16 rounded-2xl border border-black/5 shadow-sm ${getIconColor(index)} overflow-hidden`}"
		/>
	`;
}

enum IconType {
	uri,
	material,
	dashboard,
	selfhst,
}

interface IIconBaseProps {
	icon: string;
	iconColor?: string;
	iconBG?: string;
	iconBubble?: boolean;
	iconBubblePadding?: boolean;
	iconAspect?: IconAspect;
}

function IconBase(props: IIconBaseProps) {
	let { icon, iconBG, iconBubble, iconBubblePadding, iconColor, iconAspect = "square" } = props;

	let iconType: IconType = IconType.uri;

	if (icon.startsWith("http") || icon.startsWith("/")) {
		iconType = IconType.uri;
	} else if (icon.startsWith("mdi-")) {
		iconType = IconType.material;
	} else if (icon.startsWith("selfhst-")) {
		iconType = IconType.selfhst;
	} else {
		iconType = IconType.dashboard;
	}

	// Everyone starts the same size
	let bubbleClassName = "flex items-center justify-center overflow-hidden bg-contain";
	let iconWrapperWidthHeightClassName = "";
	let iconItselfWidthHeightClassName = "";

	switch (iconAspect) {
		case "width":
			iconItselfWidthHeightClassName = "w-16";
			iconWrapperWidthHeightClassName += " w-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "w-14";
			}
			break;
		case "height":
			iconItselfWidthHeightClassName = "h-16";
			iconWrapperWidthHeightClassName += " h-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "h-14";
			}
			break;
		case "square":
		default:
			iconItselfWidthHeightClassName = "w-16 h-16 object-contain";
			iconWrapperWidthHeightClassName += " w-16 h-16";

			if (iconBubblePadding) {
				iconItselfWidthHeightClassName = "w-14 h-14 object-contain";
			}
			break;
	}

	bubbleClassName += iconWrapperWidthHeightClassName;

	if (is.null(iconBubble) || iconBubble !== false) {
		bubbleClassName += " rounded-2xl border border-black/5 shadow-sm";
	}

	const bubbleStyle: string[] = [];

	switch (iconType) {
		case IconType.uri:
		case IconType.dashboard:
		case IconType.selfhst:
			// Default to bubble and no background for URI, Dashboard and selfhst icons
			if (!is.null(iconBG)) {
				if (iconBG?.startsWith("#")) {
					bubbleStyle.push(`background-color: ${iconBG}`);
				} else {
					bubbleClassName += ` bg-${iconBG}`;
				}
			}
			break;
		case IconType.material:
			// Material icons get a color and a background by default, then an optional bubble
			if (is.null(iconBG)) {
				bubbleClassName += ` bg-slate-200 dark:bg-gray-900`;
			} else {
				if (iconBG?.startsWith("#")) {
					bubbleStyle.push(`background-color: ${iconBG}`);
				} else {
					bubbleClassName += ` bg-${iconBG}`;
				}
			}

			if (is.null(iconColor)) {
				iconColor = "black dark:bg-white";
			}
			break;
	}

	const mdiIconStyle: string[] = [];
	let mdiIconColorFull = "bg-" + iconColor;

	if (!is.null(iconColor) && iconColor?.startsWith("#")) {
		mdiIconColorFull = "";
		mdiIconStyle.push(`background-color: ${iconColor}`);
	}

	switch (iconType) {
		case IconType.uri:
			return `<span class="${bubbleClassName}" style="${unwrapStyles(
				bubbleStyle
			)}"><img src="${icon}" alt="" class="${iconItselfWidthHeightClassName || ""}" /></span>`;
		case IconType.dashboard:
			icon = icon.replace(".png", "").replace(".jpg", "").replace(".svg", "");
			return `<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<img
					src=${`https://cdn.jsdelivr.net/gh/walkxcode/dashboard-icons/png/${icon}.png`}
					alt=""
					class="${iconItselfWidthHeightClassName || ""}"
				/>
			</span>`;
		case IconType.selfhst:
			icon = icon.replace("selfhst-", "").replace(".png", "").replace(".svg", "");
			return `<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<img
					src=${`https://cdn.jsdelivr.net/gh/selfhst/icons/png/${icon}.png`}
					alt=""
					class="${iconItselfWidthHeightClassName || ""}"
				/>
			</span>`;
		case IconType.material:
			icon = icon.replace("mdi-", "").replace(".svg", "");
			return `
			<span class="${bubbleClassName}" style="${unwrapStyles(bubbleStyle)}">
				<span>
					<span
						class="flex items-center justify-center ${iconItselfWidthHeightClassName} ${mdiIconColorFull} overflow-hidden"
						style="${unwrapStyles(
							mdiIconStyle.concat([
								`mask: url(https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/${icon}.svg) no-repeat center / contain`,
								`webkit-mask: url(https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/${icon}.svg) no-repeat center / contain`,
							])
						)}"
					></span>
				</span>
				</span>
			`;
	}
}

function unwrapStyles(styles: string[]): string {
	if (is.null(styles)) {
		return "";
	}

	return styles.join(";");
}
