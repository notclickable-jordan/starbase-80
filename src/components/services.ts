import { is } from "../shared/is";
import { IService } from "../shared/types";
import { Anchor } from "./anchor";
import { Icon } from "./icon";

interface IServicesProps {
	services: IService[];
	categoryBubblePadding?: boolean;
}

export const Services = function (props: IServicesProps) {
	const { services, categoryBubblePadding } = props;

	return `
		<ul class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-1 lg:gap-2 lg:gap-y-4">
			${services.map((service, index) => Service({ index, service, categoryBubblePadding })).join("")}
		</ul>
	`;
};

interface IServiceProps {
	service: IService;
	index: number;
	categoryBubblePadding?: boolean;
}

function Service(props: IServiceProps) {
	const { service, index, categoryBubblePadding } = props;

	const {
		name,
		uri,
		description,
		icon,
		iconLight,
		iconDark,
		iconBG,
		iconBubble,
		iconBubblePadding,
		iconColor,
		iconAspect,
		newWindow,
	} = service;

	return `
		<li class="p-4">
		${Anchor({
			uri,
			newWindow,
			className: "flex gap-4",
			children: `${
				!is.null(icon)
					? `<span class="flex-shrink-0 flex">${Icon({
							name,
							icon,
							iconLight,
							iconDark,
							uri,
							index,
							iconColor,
							iconBG,
							iconBubble,
							iconAspect,
							newWindow,
							categoryBubblePadding,
							iconBubblePadding,
						})}</span>`
					: ``
			}
		<div>
			<h3 class="text-lg mt-1 font-semibold line-clamp-1">${name}</h3>
			${!is.null(description) ? `<p class="text-sm text-black/50 dark:text-white/50 line-clamp-1">${description}</p>` : ``}
		</div>`,
		})}
			
		</li>
	`;
}
