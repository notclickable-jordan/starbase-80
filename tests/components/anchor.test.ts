import { Anchor } from "../../src/components/anchor";

describe("Anchor Component", () => {
	const testColors = ["blue", "rose", "green", "red", "yellow"];
	const longString =
		"This is a very long string that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the anchor element";

	describe("Required props only", () => {
		test("should render with just uri", () => {
			const result = Anchor({ uri: "https://example.com" });
			expect(result).toMatchSnapshot();
		});
	});

	describe("With title variations", () => {
		test("should render with short title", () => {
			const result = Anchor({
				uri: "https://example.com",
				title: "Short Title",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with long title", () => {
			const result = Anchor({
				uri: "https://example.com",
				title: longString,
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe("With className variations", () => {
		testColors.forEach((color, index) => {
			test(`should render with ${color} className`, () => {
				const result = Anchor({
					uri: "https://example.com",
					className: `text-${color}-500 bg-${color}-100 hover:bg-${color}-200`,
				});
				expect(result).toMatchSnapshot();
			});
		});

		test("should render with long className", () => {
			const result = Anchor({
				uri: "https://example.com",
				className:
					"flex items-center justify-center px-4 py-2 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500",
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe("With children variations", () => {
		test("should render with simple text children", () => {
			const result = Anchor({
				uri: "https://example.com",
				children: "Click me!",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with HTML children", () => {
			const result = Anchor({
				uri: "https://example.com",
				children: "Launch App",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with long children content", () => {
			const result = Anchor({
				uri: "https://example.com",
				children: longString,
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe("With newWindow variations", () => {
		test("should render with newWindow explicitly true", () => {
			const result = Anchor({
				uri: "https://example.com",
				newWindow: true,
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with newWindow explicitly false", () => {
			const result = Anchor({
				uri: "https://example.com",
				newWindow: false,
			});
			expect(result).toMatchSnapshot();
		});
	});

	describe("Combined prop variations", () => {
		test("should render with some required and optional props", () => {
			const result = Anchor({
				uri: "https://example.com",
				title: "Example Site",
				children: "Visit Example",
			});
			expect(result).toMatchSnapshot();
		});

		test("should render with all props", () => {
			const result = Anchor({
				uri: "https://example.com",
				title: "Complete Example",
				className: "btn btn-primary",
				children: '<i class="icon-external"></i> External Link',
				newWindow: true,
			});
			expect(result).toMatchSnapshot();
		});

		testColors.forEach((color, index) => {
			test(`should render with all props and ${color} theme`, () => {
				const result = Anchor({
					uri: `https://${color}.example.com`,
					title: `${color.charAt(0).toUpperCase() + color.slice(1)} themed link`,
					className: `text-${color}-600 hover:text-${color}-800 underline`,
					children: `${color.charAt(0).toUpperCase() + color.slice(1)} Link`,
					newWindow: index % 2 === 0,
				});
				expect(result).toMatchSnapshot();
			});
		});

		test("should render with all props using long strings", () => {
			const result = Anchor({
				uri: "https://very-long-domain-name-for-testing-purposes.example.com/with/a/very/long/path/that/might/cause/issues",
				title: longString,
				className:
					"inline-flex items-center px-6 py-3 border border-transparent text-base font-medium rounded-md shadow-sm text-white bg-gradient-to-r from-purple-600 to-indigo-600 hover:from-purple-700 hover:to-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 transition-all duration-200",
				children: `<div class="flex items-center space-x-2"><span>${longString}</span></div>`,
				newWindow: true,
			});
			expect(result).toMatchSnapshot();
		});
	});
});
