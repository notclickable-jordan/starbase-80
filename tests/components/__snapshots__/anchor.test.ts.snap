// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Anchor Component Combined prop variations should render with all props 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="Complete Example" class="btn btn-primary">
				<i class="icon-external"></i> External Link
			</a>
		"
`;

exports[`Anchor Component Combined prop variations should render with all props and blue theme 1`] = `
"
			<a href="https://blue.example.com" target="_blank" rel="noopener noreferrer" title="Blue themed link" class="text-blue-600 hover:text-blue-800 underline">
				ðŸŽ¨ Blue Link
			</a>
		"
`;

exports[`Anchor Component Combined prop variations should render with all props and green theme 1`] = `
"
			<a href="https://green.example.com" target="_blank" rel="noopener noreferrer" title="Green themed link" class="text-green-600 hover:text-green-800 underline">
				ðŸŽ¨ Green Link
			</a>
		"
`;

exports[`Anchor Component Combined prop variations should render with all props and red theme 1`] = `
"
		<a href="https://red.example.com" title="Red themed link" class="text-red-600 hover:text-red-800 underline">
			ðŸŽ¨ Red Link
		</a>
	"
`;

exports[`Anchor Component Combined prop variations should render with all props and rose theme 1`] = `
"
		<a href="https://rose.example.com" title="Rose themed link" class="text-rose-600 hover:text-rose-800 underline">
			ðŸŽ¨ Rose Link
		</a>
	"
`;

exports[`Anchor Component Combined prop variations should render with all props and yellow theme 1`] = `
"
			<a href="https://yellow.example.com" target="_blank" rel="noopener noreferrer" title="Yellow themed link" class="text-yellow-600 hover:text-yellow-800 underline">
				ðŸŽ¨ Yellow Link
			</a>
		"
`;

exports[`Anchor Component Combined prop variations should render with all props using long strings 1`] = `
"
			<a href="https://very-long-domain-name-for-testing-purposes.example.com/with/a/very/long/path/that/might/cause/issues" target="_blank" rel="noopener noreferrer" title="This is a very long string that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the anchor element" class="inline-flex items-center px-6 py-3 border border-transparent text-base font-medium rounded-md shadow-sm text-white bg-gradient-to-r from-purple-600 to-indigo-600 hover:from-purple-700 hover:to-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 transition-all duration-200">
				<div class="flex items-center space-x-2"><span class="text-xl">ðŸ”—</span><span>This is a very long string that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the anchor element</span></div>
			</a>
		"
`;

exports[`Anchor Component Combined prop variations should render with some required and optional props 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="Example Site" class="">
				Visit Example
			</a>
		"
`;

exports[`Anchor Component Required props only should render with just uri 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="">
				undefined
			</a>
		"
`;

exports[`Anchor Component With children variations should render with HTML children 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="">
				<span class="icon">ðŸš€</span> Launch App
			</a>
		"
`;

exports[`Anchor Component With children variations should render with long children content 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="">
				This is a very long string that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the anchor element
			</a>
		"
`;

exports[`Anchor Component With children variations should render with simple text children 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="">
				Click me!
			</a>
		"
`;

exports[`Anchor Component With className variations should render with blue className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="text-blue-500 bg-blue-100 hover:bg-blue-200">
				undefined
			</a>
		"
`;

exports[`Anchor Component With className variations should render with green className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="text-green-500 bg-green-100 hover:bg-green-200">
				undefined
			</a>
		"
`;

exports[`Anchor Component With className variations should render with long className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="flex items-center justify-center px-4 py-2 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
				undefined
			</a>
		"
`;

exports[`Anchor Component With className variations should render with red className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="text-red-500 bg-red-100 hover:bg-red-200">
				undefined
			</a>
		"
`;

exports[`Anchor Component With className variations should render with rose className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="text-rose-500 bg-rose-100 hover:bg-rose-200">
				undefined
			</a>
		"
`;

exports[`Anchor Component With className variations should render with yellow className 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="text-yellow-500 bg-yellow-100 hover:bg-yellow-200">
				undefined
			</a>
		"
`;

exports[`Anchor Component With newWindow variations should render with newWindow explicitly false 1`] = `
"
		<a href="https://example.com" title="" class="">
			undefined
		</a>
	"
`;

exports[`Anchor Component With newWindow variations should render with newWindow explicitly true 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="" class="">
				undefined
			</a>
		"
`;

exports[`Anchor Component With title variations should render with long title 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="This is a very long string that tests how the component handles extensive text content and ensures that it properly renders without breaking the layout or functionality of the anchor element" class="">
				undefined
			</a>
		"
`;

exports[`Anchor Component With title variations should render with short title 1`] = `
"
			<a href="https://example.com" target="_blank" rel="noopener noreferrer" title="Short Title" class="">
				undefined
			</a>
		"
`;
